server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceComputer.java

```diff
@@ -77,6 +80,8 @@ public class DesiredBalanceComputer {
     private long lastConvergedTimeMillis;
     private long lastNotConvergedLogMessageTimeMillis;
     private Level convergenceLogMsgLevel;
+    private ClusterInfo lastClusterInfo;
+    private Map<ShardForSimulation, ShardRouting> shardsStartedByAllocate;
```

Comment:
Can we see this as input instead? I think I prefer to track this outside this class rather than being this stateful here.

Reply:
It could even reside on or be accessible through the `ClusterInfo` object?